storage/src/main/java/tech/pegasys/teku/storage/protoarray/ProtoArray.java

```diff
@@ -266,8 +273,21 @@ public void markNodeInvalid(final Bytes32 blockRoot, final Optional<Bytes32> lat
     final int index;
     final ProtoNode node;
     if (latestValidHash.isPresent()) {
-      final Optional<Integer> maybeFirstInvalidNodeIndex =
-          findFirstInvalidNodeIndex(maybeIndex.get(), latestValidHash.get());
+      final Optional<Integer> maybeFirstInvalidNodeIndex;
+      if (verifiedInvalidTransition) {
+        maybeFirstInvalidNodeIndex =
+            findFirstInvalidNodeIndex(maybeIndex.get(), latestValidHash.get());
+      } else {
+        maybeFirstInvalidNodeIndex =
+            verifyInvalidIndex(maybeIndex.get(), latestValidHash.get())
+                .flatMap(
+                    invalidIndex -> findFirstInvalidNodeIndex(invalidIndex, latestValidHash.get()));
```

Comment:
I think it would be clearer with an if rather than using an Optional. So I think I'd do something like:

```java
  public void markNodeInvalid(
      final Bytes32 blockRoot,
      final Optional<Bytes32> latestValidHash,
      final boolean verifiedInvalidTransition) {
    if (!verifiedInvalidTransition && latestValidHash.isEmpty()) {
      // Couldn't find invalid chain segment with lack of data
      return;
    }
    final Optional<Integer> maybeIndex = indices.get(blockRoot);
    if (maybeIndex.isEmpty()) {
      LOG.debug("Couldn't update status for block {} because it was unknown", blockRoot);
      return;
    }
    final int index;
    final ProtoNode node;
    if (latestValidHash.isPresent()) {
      final Optional<Integer> maybeFirstInvalidNodeIndex =
          findFirstInvalidNodeIndex(maybeIndex.get(), latestValidHash.get());
      if (!verifiedInvalidTransition) {
        if (nodeHasExecutionHash(maybeIndex.get(), latestValidHash.get())) {
          // head blockRoot is the latestValidHash so nothing to do
          return;
        }
        if (maybeFirstInvalidNodeIndex.isEmpty()) {
          // Nothing to do: latestValidHash was not found, no proof of invalid transition
          return;
        }
      }

      index = maybeFirstInvalidNodeIndex.orElse(maybeIndex.get());
      node = getNodeByIndex(index);
      // We found the latestValidHash so mark it as valid
      if (maybeFirstInvalidNodeIndex.isPresent()) {
        markNodeValid(node.getParentRoot());
      }
    } else {
      index = maybeIndex.get();
      node = getNodeByIndex(index);
    }

    node.setValidationStatus(INVALID);
    removeBlockRoot(node.getBlockRoot());
    markDescendantsAsInvalid(index);
    // Applying zero deltas causes the newly marked INVALID nodes to have their weight set to 0
    applyDeltas(new LongArrayList(Collections.nCopies(getTotalTrackedNodeCount(), 0L)));
  }

  private boolean nodeHasExecutionHash(final int nodeIndex, final Bytes32 executionHash) {
    return getNodeByIndex(nodeIndex).getExecutionBlockHash().equals(executionHash);
  }
```